refactored TextStyle to allow const construction

dunmatt opened the pull request:
A few tweaks to allow `TextStyle`s to be created as constants, eg:

    /// The default text style for widgets.
    pub const DEFAULT_TEXT_STYLE: TextStyle = TextStyle {
        font: "ProFont For Powerline",
        point_size: 11,
        fg: Color::from_rgb(0xEB, 0xDB, 0xB2),
        bg: Some(Color::from_rgb(0xEB, 0xDB, 0xB2)),
        padding: (2.0, 2.0),
    };

Note:  In order to create `Color`s in a `const` context it was necessary to switch the channel representation to integers.  As a nice side effect of doing that I also fixed a bug in `new_from_hex` caused by the ambiguous u32 representation scheme it expects, and that allowed some other simplifications.

Commit:
refactored TextStyle to allow const construction

Commit:
fixed formatting

Commit:
removed leading zeros to placate clippy

Commit:
removed no-ops to placate clippy

Commit:
removed extraneous intos to placate clippy

Commit:
minor cleanup

dunmatt commented:
Friendly ping!

sminez commented:
Hi @dunmat 
Sorry, can you provide some more details on why you are wanting to make this change? I'm not very familiar with the benefits of const but this seems to mean we are converting to f64s (with division) every time a colour is used. Unless I'm misunderstanding something...

You've also removed several tests, can you let me know why?

dunmatt commented:
The main (only?) benefit of const in this context is cleanliness, if you look at the changes to the example code you can see it in action; boilerplate calls to `into` go away, and `Color` consts start having a more appropriate type.

It's true that we're converting to f64s whenever you call `rgba`, but when I grepped the codebase it didn't look like you were doing that anywhere, so it seems like a no-brainer cleanup.

As for the removal of those tests, those tests were codifying unreasonable behavior.  I think you (or whoever wrote the line `/// Create a new Color from a hex encoded u32: 0xRRGGBB or 0xRRGGBBAA`) were a bit confused about the relationship between hexadecimal int literals and strings representing the same.  I say that because that docstring describes an ambiguous encoding scheme.  Take the first test I removed, for example:  `Color::from(0x00000000)`, the problem is `0x0000_0000 == 0x00_0000` (they both also `== 0` and `== 0b00` and `== 0o00`), and so it is impossible to differentiate if the user is asking for black or invisible black.  Why anyone would ever ask for invisible black is beyond me, so I deleted the test that asserted it as correct behavior.  I also filed #76 to address the ambiguity problem more directly, let me know if you'd like me to make that change as well.

sminez commented:
Ah ok, that makes a bit more sense, thanks 🙂 

In terms of converting to f64s, it happens [here](https://github.com/sminez/penrose/blob/588c7771a31681d215e551859b84be9f0a4d08bd/src/draw/mod.rs#L339) when setting the color for a DrawContext which is then used for things like [drawing rectangles](https://github.com/sminez/penrose/blob/588c7771a31681d215e551859b84be9f0a4d08bd/src/draw/mod.rs#L369) or [rendering a layout](https://github.com/sminez/penrose/blob/588c7771a31681d215e551859b84be9f0a4d08bd/src/draw/mod.rs#L384). Both of which are used on every render call to the status bar. I'd really rather not do that float conversion every time the color is set: is there any way to keep the const stuff but with f64s instead of u8s?

dunmatt commented:
Is there some reason you wouldn't want const constructors?  If it's a problem I can hack around it using `lazy_static`, I'd rather not block on this (I have other changes waiting)

dunmatt commented:
Not on stable.  You can do it in nightly, but that's probably not worth it.  I can also hack around the issue via `lazy_static` if that's preferable.

dunmatt commented:
I've refactored to use `lazy_static` on my end, and the only change I need from this PR is now also covered by #77 , so I'm going to close this one.
